File: lasso_minibatch.cpp
#include <iostream>
#include <Eigen/Dense>
#include <random>
#include <chrono>

using namespace Eigen;

double soft_thresholding(double z, double gamma) {
    return std::copysign(1.0, z) * std::max(std::abs(z) - gamma, 0.0);
}

VectorXd lasso_coordinate_descent(const MatrixXd& X, const VectorXd& y, double lambda=0.1, int num_epoch=200, double tol=1e-4, double lr=1.0) {
    int n_samples = X.rows();
    int n_features = X.cols();
    //VectorXd beta = VectorXd::Random(n_features);
    VectorXd beta = VectorXd::Zero(n_features);
    int batch_size = 20;
    std::default_random_engine generator;
    
    for (int epoch = 0; epoch < num_epoch; ++epoch) {
        std::vector<int> indices(n_samples);
        for (int i = 0; i < n_samples; ++i) indices[i] = i;
        std::shuffle(indices.begin(), indices.end(), generator);
        
        for (int i = 0; i < n_samples; i += batch_size) {
            VectorXd beta_old = beta;
            int current_batch_size = std::min(batch_size, n_samples - i);
            MatrixXd X_batch(current_batch_size, n_features);
            VectorXd y_batch(current_batch_size);
            for (int b = 0; b < current_batch_size; ++b) {
                int idx = indices[i + b];
                X_batch.row(b) = X.row(idx);
                y_batch(b) = y(idx);
            }
            
            VectorXd residual = y_batch - X_batch * beta;
            double max_delta = 0.0;
            
            for (int j = 0; j < n_features; ++j) {
                //double beta_old_j = beta(j);
                double gradient = -X_batch.col(j).dot(residual) / n_samples;
                beta(j) = beta_old(j) - lr * gradient;
                beta(j) = soft_thresholding(beta(j), lambda);
                double delta = beta(j) - beta_old(j);
                residual -= X_batch.col(j) * delta;
                if (std::abs(delta) > max_delta) {
                    max_delta = std::abs(delta);
                }
            }
            
            if (max_delta < tol) {
                std::cout << "Converged in " << epoch + 1 << " epochs." << std::endl;
                return beta;
            }
        }
    }
    std::cout << "Reached maximum epochs: " << num_epoch << std::endl;
    return beta;
}

int main() {
    std::srand(0);
    int n = 100, p = 1000;
    MatrixXd X(n, p);
    VectorXd y(n);
    VectorXd true_beta = VectorXd::Zero(p);
    true_beta(0) = 2.0;
    true_beta(1) = -1.0;
    true_beta(2) = 0.5;
    true_beta(4) = 3.0;
    
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < p; ++j) {
            X(i, j) = std::rand() / double(RAND_MAX);
        }
        y(i) = X.row(i).dot(true_beta) + (std::rand() / double(RAND_MAX)) * 0.5;
    }
    
    VectorXd X_mean = X.colwise().mean();
    X = X.rowwise() - X_mean.transpose();
    VectorXd X_std = ((X.array().square().colwise().sum() / (n - 1)).sqrt());
    X = X.array().rowwise() / X_std.transpose().array();
    y = y.array() - y.mean();
    
    auto start = std::chrono::high_resolution_clock::now();
    VectorXd beta_est = lasso_coordinate_descent(X, y);
    auto end = std::chrono::high_resolution_clock::now();
    
    std::cout << "Estimated coefficients:\n" << beta_est.transpose() << std::endl;
    std::cout << "\nTime: " 
              << std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count()
              << " ms" << std::endl;
    
    return 0;
}

File: lasso_minibatch.py
import time
import numpy as np


def soft_thresholding(z, gamma):
    return np.sign(z) * np.maximum(np.abs(z) - gamma, 0.0)


def lasso_coordinate_descent(X, y, lambda_=0.1, num_epoch=200, tol=1e-4, lr=1.0):
    n_samples, n_features = X.shape
    beta = np.random.random(n_features)
    batch_size = 20

    for epoch in range(num_epoch):
        #beta_old = beta.copy()
        row_number = np.random.permutation(n_samples)
        for i in range(0, n_samples, batch_size):
            beta_old = beta.copy() # 
            batch_pick = row_number[i : i + batch_size]
            X_random = X[batch_pick]
            residual_random = y[batch_pick] - X[batch_pick] @ beta
            max_delta = 0

            for j in range(n_features):
                gk = -np.dot(X_random[:, j], residual_random) / n_samples 
                beta[j] = beta_old[j] - lr * gk
                beta[j] = soft_thresholding(beta[j], lambda_)
                delta = beta[j] - beta_old[j]
                if np.abs(delta) > max_delta:          
                    max_delta = np.abs(delta)         
                residual_random -= X_random[:, j] * delta

            if max_delta < tol:          # change this convergence checking to the same as glmnet
                print(f"Converged in {epoch + 1} epochs.")
                return beta
        if epoch >= num_epoch - 3:
            print(f"Epoch {epoch + 1} beta[:10]:", np.round(beta[:10], 4))

    return beta


# Simulated data
np.random.seed(0)
n, p = 100, 1000
X = np.random.randn(n, p)
true_beta = np.zeros(p)
true_beta[:5] = [2, -1, 0.5, 0, 3]  # sparse true coefficients
y = X @ true_beta + np.random.randn(n) * 0.5

# Standardize X and center y
X -= X.mean(axis=0)
X /= X.std(axis=0)
y -= y.mean()

# Fit LASSO
start = time.time()
beta_est = lasso_coordinate_descent(X, y)
end = time.time()

print("Estimated coefficients:\n", np.round(beta_est, 3))
print("Time: ", (end - start) * 1000, "ms")